### ISSUE TITLE
Hydra channel for dhall-haskell failure

### ISSUE THREAD

**Issue by jpotier** [dhall-lang/dhall-lang#1006]
https://hydra.dhall-lang.org/jobset/dhall-haskell/master/channel/latest

returns an error:

```
DBIx::Class::Storage::DBI::_dbh_execute(): DBI Exception: DBD::Pg::st execute failed: ERROR: more than one row returned by a subquery used as an expression [for Statement "SELECT me.id, me.finished, me.timestamp, me.stoptime, me.project, me.jobset, me.job, me.nixname, me.system, me.buildstatus, me.releasename, me.drvpath, me.description, me.homepage, buildoutputs.path, buildoutputs.name FROM ( select *
from
(select
(select max(b.id) from builds b
where
project = activeJobs.project and jobset = activeJobs.jobset
and job = activeJobs.job and system = activeJobs.system
and finished = 1 and buildstatus = 0
) as id
from (select distinct project, jobset, job, system from Builds where isCurrent = 1 and jobset_id = (select id from jobsets j where j.name = ?)) as activeJobs
) as latest
join Builds using (id)
) me LEFT JOIN buildoutputs buildoutputs ON buildoutputs.build = me.id WHERE ( exists (select 1 from buildproducts where build = me.id and type = 'nix-build') ) ORDER BY me.id, buildoutputs.name" with ParamValues: 1='master'] at /nix/store/8k8qq24y86jyzx93s8w37wm5ncmiips5-hydra-2020-04-16/libexec/hydra/lib/Hydra/Base/Controller/NixChannel.pm line 20
```

My nix configuration currently picks https://hydra.dhall-lang.org/jobset/dhall-haskell/master/channel/latest/nixexprs.tar.bz2, which throws an HTTP 500, preventing my system to build :D 

Am I wrong to expect that channel to exist and return 200?

**Comment by philandstuff** [comment]
I've done a bit of digging, and though I haven't solved it I thought I'd dump my understanding here in case it's helpful for someone else:

## The problematic bit of the query

The query in OP is a bit big to dig through, but thankfully there's a tiny subquery which highlights the problem:

```sql
select distinct project, jobset, job, system from Builds where isCurrent = 1 and jobset_id = (select id from jobsets j where j.name = 'master');
```

If I run this at the `psql` console, this query fails with the same `ERROR:  more than one row returned by a subquery used as an expression` error we get in the web interface.

If I run the subquery `select id from jobsets j where j.name = 'master'` I get 4 results: `dhall-kubernetes/master`, `dhall-haskell/master`, `dhall-to-cabal/master` and `dhall-lang/master`.  This makes sense because the primary key for jobset isn't the `name` field, it is the `(project,name)` fields.  But I suspect the desired query was to select only `dhall-haskell/master` and not the other 3, since the URL is in the context of the `dhall-haskell` jobset.

## What might be causing it?

The [recent hydra migration](https://github.com/NixOS/hydra/pull/710) is changing this stuff so that there will be an id for jobsets.  It seems that, for an intervening time, we had some glue queries that converted old-style `name`-based queries into new-style `id`-based queries, and this looks like one of them: we've got a jobset `name` and we want to get its `id`.  The problem is that, because we have jobsets with the same name in different projects, we're not getting distinct results.

Looking at the contents of the database tables, it seems that we have completed step 8 of the migration plan in issue NixOS/hydra#710 - we have `jobset_id` columns on the `build` and `job` tables, and they have been populated, and they have been modified to be `not null`.  So it's strange to me that we're still executing `name`-based queries when we could be running a version of the code which uses entirely `id`-based queries which should avoid these multiple results problems.

## A smoking gun

Commit NixOS/hydra@548fd8eadd88 looks suspiciously like it fixes the problem we're seeing - the commit message suggests there was a problem in the original "part 2" of the migration, and mentions our failure mode "specifically breaks when two projects have a jobset of the same name."  I don't know where in nixpkgs that commit will be pulled in, but my gut feel is that if we now bump from `hydra-migration` to `hydra-unstable` it should all Just Work?

It's past 10pm local time for me, though, so i'm not going to try to change the package version until the morning.

## Next steps

Maybe just try upgrading from `hydra-migration` to `hydra-unstable`?

Maybe try to understand which package will include commit NixOS/hydra@548fd8eadd88 for us, which I think will fix our nix channels?

**Comment by Gabriel439** [comment]
@philandstuff: Thanks a lot for investigating this!  I think your diagnosis is correct

Also, there is an easy way to check if your change has the intended effect, using `nix-diff`:

https://github.com/Gabriel439/nix-diff

Just run `nix-instantiate --attr machine release.nix` before and after your changes and then use `nix-diff` to compare the changes to verify that the things you expect to change are indeed changing.

**Comment by philandstuff** [comment]
Fixed by #1009.

**Comment by Gabriel439** [comment]
@philandstuff: Thank you again for fixing this!  I really appreciate it 🙂
